src/tools/interop/idt/capture/platform/android/android.py

```diff
@@ -0,0 +1,220 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+import ipaddress
+import os
+import time
+import typing
+
+from capture.base import PlatformLogStreamer
+from capture.file_utils import create_standard_log_name
+from capture.shell_utils import BashRunner
+
+
+class Android(PlatformLogStreamer):
+    """
+    Class that supports:
+    - Running synchronous adb commands
+    - Maintaining a singleton logcat stream
+    - Maintaining a singleton screen recording
+    """
+
+    def __init__(self, artifact_dir: str) -> None:
+
+        self.artifact_dir = artifact_dir
+
+        self.device_id: str | None = None
+        self.adb_devices: typing.Dict[str, bool] | None = None
+        self._authorize_adb()
+
+        # TODO: Set a larger buffer size
+        # TODO: More efficient buffer for analysis than the file
+        self.logcat_output_path = os.path.join(
+            self.artifact_dir, create_standard_log_name(
+                'logcat', 'txt'))
+        self.logcat_proc: BashRunner | None = None
+        self.logcat_command = f'adb -s {self.device_id} logcat -T 1 >> {self.logcat_output_path}'
```

Comment:
Here is a logcat script I wrote a few years ago to try and follow the process ids...

```python
import argparse
import re
import subprocess

from subprocess import PIPE


import colorama as colorama
from colorama import Back, Fore, Style
from pydantic import BaseModel


# Proc started.
# 08-21 15:14:23.880   590   610 I ActivityManager: Start proc 4426:com.amazon.venezia:sync/u0a22096 for service com.amazon.venezia/com.amazon.mas.client.notifications.NotificationSyncAdapter$NotificationSyncService
PID_START_GROUPS = ("pid", "procname", "uid", "hostingname")
PID_START_PATTERN = re.compile(
    r"^Start proc (?P<%s>\d+):(?P<%s>[a-zA-Z0-9._:]+)/(?P<%s>[a-z0-9]+) for (?P<%s>.*)$"
    % PID_START_GROUPS
)

# Proc killed.
# 08-21 15:29:07.975   582  1266 I ActivityManager: Killing 4299:com.amazon.sync.service/u0a22112 (adj 12): empty #13
PID_KILL_GROUPS = ("pid", "procname", "uid", "adj", "reason")
PID_KILL_PATTERN = re.compile(
    r"^Killing (?P<%s>\d+):(?P<%s>[a-zA-Z0-9._:]+)/(?P<%s>\w+) \(adj (?P<%s>.*)\): (?P<%s>.*)$"
    % PID_KILL_GROUPS
)

# Proc died.
# 08-21 15:30:14.222   582  1051 I ActivityManager: Process com.amazon.kor.demo (pid 7340) has died
PID_DEATH_GROUPS = ("procname", "pid")
PID_DEATH_PATTERN = re.compile(
    r"^Process (?P<%s>[a-zA-Z0-9._:]+) \(pid (?P<%s>\d+)\) has died.?$"
    % PID_DEATH_GROUPS
)


PROGRAM = ["shell", "ps"]
OPTIONS = []

# USER           PID  PPID     VSZ    RSS WCHAN            ADDR S NAME
# root             1     0   51276   6628 SyS_epoll_wait      0 S init
PS_GROUP_NAMES = ("user", "pid", "ppid", "vsize", "rss", "wchan", "addr", "s", "name")

PS_PATTERN = re.compile(
    r"(?P<user>\w+)\s+(?P<pid>\w+)\s+(?P<ppid>\w+)\s+(?P<vsize>\w+)\s+(?P<rss>\w+)\s+(?P<wchan>\w+)\s+(?P<addr>\w+)\s+(?P<s>\w+)\s+(?P<name>[^ ]+)"
)


class PsInfo(BaseModel):
    pid: int = None
    name: str = None


def process_adb_shell_ps_output_line(ps_output_line):
    fields = ps_output_line.split()

    try:
        psinfo: PsInfo = PsInfo(
            pid=int(fields[1]),
            name=str(fields[8]),
        )
    except Exception as e:
        return None

    return psinfo


def execute_adb_shell_ps(adb_command):
    return adblib.generate_shell_output(
        adb_command, PROGRAM, OPTIONS, process_adb_shell_ps_output_line
    )


# Execute a shell command and return the results as an iterable.
def generate_shell_output(
    adb_command, shell_command, shell_command_options, transform_func
):
    command = adb_command + shell_command + shell_command_options
    process = subprocess.Popen(
        command, stdin=PIPE, stdout=PIPE, stderr=PIPE, universal_newlines=True
    )

    # While no return status, process stdout lines.
    while process.poll() is None:
        try:
            line = process.stdout.readline()
        except KeyboardInterrupt:
            break
        except Exception as e:
            continue

        output = transform_func(line)
        if output is not None:
            yield output


ADB_COMMAND = ["adb"]

# Example: 02-02 22:23:40.019  1085  1363 E tag: message
LOGCAT_GROUPS = ("datetime", "pid", "ppid", "verbosity", "tag", "message")

LOGCAT_PATTERN = re.compile(
    r"^(?P<%s>\d{2}-\d{2} \d{2}:\d{2}:\d{2}.\d{3})\s+"
    r"(?P<%s>\d+)\s+"
    r"(?P<%s>\d+)\s+"
    r"(?P<%s>.)\s+"
    r"(?P<%s>.+?):\s+"
    r"(?P<%s>.+)$" % LOGCAT_GROUPS
)


def isdecimal(x):
    try:
        int(x)
        return True
    except:
        return False


def main():
    colorama.init()

    parser = argparse.ArgumentParser(description="adb logcat enhanced")
    parser.add_argument(
        "accept_packages", nargs="*", help="Packages to accept while filtering"
    )
    parser.add_argument(
        "-d",
        "--drop",
        nargs="*",
        dest="drop_packages",
        help="Packages to drop while filtering",
    )
    parser.add_argument(
        "-s",
        "--serial",
        dest="device_serial",
        help="Device serial number (adb -s option)",
    )
    args = parser.parse_args()

    if args.device_serial:
        ADB_COMMAND.extend(["-s", args.device_serial])

    ps_infos = psinfo.execute_adb_shell_ps(ADB_COMMAND)

    # Create lookup map for process info, using pid as key.
    ps_info_dict = {}
    for ps_info in ps_infos:
        ps_info_dict[ps_info.pid] = ps_info

    # Get logcat iterable.
    adb_logcat_outputs = process_adb_logcat_output()

    # For each line in the logcat output.
    for adb_logcat_output in adb_logcat_outputs:
        # If new process is starting
        pid_start_match = procmon.PID_START_PATTERN.match(adb_logcat_output["message"])
        if pid_start_match is not None:
            ps_info = psinfo.PsInfo(
                pid=pid_start_match.group("pid"), name=pid_start_match.group("procname")
            )
            ps_info_dict[ps_info.pid] = ps_info

        m = procmon.PID_KILL_PATTERN.match(adb_logcat_output["message"])
        if m is not None and isdecimal(m.group(1)):
            killed_pid = int(m.group(1))
            ps_info_dict.pop(killed_pid, None)

        m = procmon.PID_DEATH_PATTERN.match(adb_logcat_output["message"])
        if m is not None and isdecimal(m.group(2)):
            death_pid = int(m.group(2))
            ps_info_dict.pop(death_pid, None)

        # If packages to drop defined, drop them now.
        if args.drop_packages and len(args.drop_packages) > 0:
            pid = int(adb_logcat_output["pid"])
            if pid in ps_info_dict and ps_info_dict[pid].name in args.drop_packages:
                continue

        # If packages defined, filter to only show those lines now.
        if args.accept_packages and len(args.accept_packages) > 0:
            pid = int(adb_logcat_output["pid"])
            if (
                pid not in ps_info_dict
                or ps_info_dict[pid].name not in args.accept_packages
            ):
                continue

        logcat_output = format_logcat_output(ps_info_dict, adb_logcat_output)
        print(logcat_output)


def process_adb_logcat_output():
    program = ["logcat"]
    options = ["-b", "all", "-v", "threadtime"]

    adb_logcat_outputs = adblib.generate_shell_output(
        ADB_COMMAND, program, options, parse_logline
    )
    for adb_logcat_output in adb_logcat_outputs:
        if adb_logcat_output is not None:
            yield adb_logcat_output


def format_logcat_output(ps_info_dict, adb_logcat_output):
    pid = int(adb_logcat_output["pid"])
    name = ps_info_dict[pid].name if pid in ps_info_dict else f"Unknown({pid})"

    linebuf = (
        Fore.YELLOW
        + Style.BRIGHT
        + adb_logcat_output["datetime"]
        + Style.RESET_ALL
        + " "
    )
    linebuf += (
        Fore.YELLOW + Style.BRIGHT + adb_logcat_output["pid"] + Style.RESET_ALL + " "
    )
    linebuf += (
        Fore.YELLOW + Style.BRIGHT + adb_logcat_output["ppid"] + Style.RESET_ALL + " "
    )

    message = f'{adb_logcat_output["verbosity"]} {name} {adb_logcat_output["tag"]} {adb_logcat_output["message"]}'

    if adb_logcat_output["verbosity"] == "V":
        linebuf += Fore.GREEN + Style.BRIGHT + message + Style.RESET_ALL + " "

    if adb_logcat_output["verbosity"] == "D":
        linebuf += Fore.BLUE + Style.BRIGHT + message + Style.RESET_ALL + " "

    if adb_logcat_output["verbosity"] == "I":
        linebuf += Fore.WHITE + Style.BRIGHT + message + Style.RESET_ALL + " "

    if adb_logcat_output["verbosity"] == "W":
        linebuf += Fore.MAGENTA + Style.BRIGHT + message + Style.RESET_ALL + " "

    if adb_logcat_output["verbosity"] == "E":
        linebuf += Fore.RED + Style.BRIGHT + message + Style.RESET_ALL + " "

    if adb_logcat_output["verbosity"] == "F":
        linebuf += Fore.RED + Back.RED + Style.BRIGHT + message + Style.RESET_ALL + " "

    return linebuf


def parse_logline(logline):
    pattern_match = LOGCAT_PATTERN.match(logline)

    if pattern_match is None:
        return None

    logline_info = {}
    for logcat_group in LOGCAT_GROUPS:
        logline_info[logcat_group] = pattern_match.group(logcat_group)

    return logline_info


if __name__ == "__main__":
    main()

```

Reply:
Thanks, this will come in handy!